relay_test.go

```diff
@@ -1014,3 +1014,161 @@ func TestRelayRaceCompletionAndTimeout(t *testing.T) {
+		SetRelayHost(inspector)
+
+	testutils.WithTestServer(t, opts, func(tb testing.TB, ts *testutils.TestServer) {
+		const (
+			testMethod = "echo"
+			arg2Data   = "arg2-is"
+			arg3Data   = "arg3-here"
+		)
+
+		var (
+			wantArg2Start = len(ts.ServiceName()) + len(testMethod) + 70 /*data before arg1*/
+			payloadLeft   = MaxFramePayloadSize - wantArg2Start
+		)
+
+		testutils.RegisterEcho(ts.Server(), nil)
+
+		rh.Add(ts.ServiceName(), ts.Server().PeerInfo().HostPort)
+		client := testutils.NewClient(t, nil /*opts*/)
+		defer client.Close()
+
+		tests := []struct {
+			msg               string
+			arg2Data          string
+			arg2Flush         bool
+			arg2PostFlushData string
+			noArg3            bool
+			wantEndOffset     int
+			wantHasMore       bool
+		}{
+			{
+				msg:           "all within a frame",
+				arg2Data:      arg2Data,
+				wantEndOffset: wantArg2Start + len(arg2Data),
+				wantHasMore:   false,
+			},
+			{
+				msg:           "arg2 flushed",
+				arg2Data:      arg2Data,
+				arg2Flush:     true,
+				wantEndOffset: wantArg2Start + len(arg2Data),
+				wantHasMore:   true,
+			},
+			{
+				msg:               "arg2 flushed called then write again",
+				arg2Data:          arg2Data,
+				arg2Flush:         true,
+				arg2PostFlushData: "more data",
+				wantEndOffset:     wantArg2Start + len(arg2Data),
+				wantHasMore:       true,
+			},
+			{
+				msg:           "no arg2",
+				wantEndOffset: wantArg2Start,
+				wantHasMore:   false,
+			},
+			{
+				msg:           "XL arg2 which is fragmented",
+				arg2Data:      string(make([]byte, MaxFrameSize+100)),
+				wantEndOffset: wantArg2Start + payloadLeft,
+				wantHasMore:   true,
+			},
+			{
+				msg:           "large arg2 with 3 bytes left for arg3",
+				arg2Data:      string(make([]byte, payloadLeft-3)),
+				wantEndOffset: wantArg2Start + payloadLeft - 3,
+				wantHasMore:   false,
+			},
+			{
+				msg:           "large arg2, 2 bytes left",
+				arg2Data:      string(make([]byte, payloadLeft-2)),
+				wantEndOffset: wantArg2Start + payloadLeft - 2,
+				wantHasMore:   true, // no arg3
+			},
+			{
+				msg:           "large arg2, 2 bytes left, no arg3",
+				arg2Data:      string(make([]byte, payloadLeft-2)),
+				wantEndOffset: wantArg2Start + payloadLeft - 2,
+				noArg3:        true,
+				wantHasMore:   true, // no arg3 and still got CALL_REQ_CONTINUE
+			},
+			{
+				msg:           "large arg2, 1 bytes left",
+				arg2Data:      string(make([]byte, payloadLeft-1)),
+				wantEndOffset: wantArg2Start + payloadLeft - 1,
+				wantHasMore:   true, // no arg3
+			},
+		}
+
+		for _, tt := range tests {
+			t.Run(tt.msg, func(t *testing.T) {
+
+				call, err := client.BeginCall(ctx, ts.HostPort(), ts.ServiceName(), testMethod, nil)
+				require.NoError(t, err, "BeginCall failed")
+				writer, err := call.Arg2Writer()
+				arg2WriteHelper := NewArgWriter(writer, err)
+				require.NoError(t, arg2WriteHelper.Write([]byte(tt.arg2Data)), "arg3 write failed")
```

Comment:
don't use the arg writer here, since it also calls `Close`. we should use the returned `Arg2Writer` directly.